fuzzer/src/main.rs

```diff
@@ -94,19 +142,75 @@ fn fuzz_one_input(data: &[u8]){
 		BTreeMap<u8, u8>,
 		BTreeSet<u32>,
 		VecDeque<u8>,
-		BinaryHeap<u32>,
+		BinaryHeapWrapper,
 		MockStruct,
 		MockEnum,
 		BitVec<BigEndian, u8>,
 		BitVec<BigEndian, u32>,
 		Duration,
+	};
+}
+
+macro_rules! fuzz_encoder {
+	() => {};
+	($( $type:ty, )*) => {
+		$(fuzz!(|data: $type| { fuzz_encode(data) });)*
+	};
+}
+
+fn fuzz_encode<T: Encode + Decode + Clone + PartialEq + std::fmt::Debug> (data: T) {
+	let original = data.clone();
+	let mut obj: &[u8] = &data.encode();
+	let decoded = <T>::decode(&mut obj);
+	if let Ok(object) = decoded {
+		if object != original {
+			println!("original object: {:?}", original);
+			println!("decoded object: {:?}", object);
+			panic!("Original object differs from decoded object")
+		}
+	} else {
+		// safe because we checked that object is not Ok
+		let e = decoded.unwrap_err();
+		println!("original object: {:?}", original);
+		println!("decoding error: {:?}", e);
+		panic!("Failed to decode the encoded object");
 	}
 }
 
+macro_rules! fuzz_encoding {
+	() => {
+		fuzz_encoder! {
+			u8,
+			u16,
+			u32,
+			u64,
+			u128,
+			Compact<u8>,
+			Compact<u16>,
+			Compact<u32>,
+			Compact<u64>,
+			Compact<u128>,
+			String,
+			Vec<u8>,
+			Vec<Vec<u8>>,
+			Option<Vec<u8>>,
+			Vec<u32>,
+			LinkedList<u8>,
+			BTreeMap<String, Vec<u8>>,
+			BTreeMap<u8, u8>,
+			BTreeSet<u32>,
+			VecDeque<u8>,
+			BinaryHeapWrapper,
+			MockStruct,
+			MockEnum,
```

Comment:
I've removed the `BitVec<BigEndian, u8>` and `BitVec<BigEndian, u32>` types here because I could not implement Arbitrary for them

Reply:
Put them in a `NewType` wrapper to implement the trait.